### PR TITLE
Warn for calls in top-level statements that could be awaited

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/64964

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:141]: It is probably better (from readability perspective) to convert this into a lambda at the only use site.

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:121]: >is SynthesizedSimpleProgramEntryPointSymbol [](http://example.com/codeflow?start=33&length=43)

It looks like the type check is redundant

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:130]: >!bag.DiagnosticBag.AsEnumerableWithoutResolution().Any(isErrorOtherThanBadAwait); [](http://example.com/codeflow?start=27&length=81)

Perhaps the logic could be easier to follow if we remove double negative (remove negation at the top, change Any to All, etc.). It looks like we we are saying that the expression CouldBeAwaited when all errors that we get here are either ```ErrorCode.ERR_BadAwaitWithoutAsync```, or ```ErrorCode.ERR_BadAwaitWithoutVoidAsyncMethod```.

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs:3843]: >AsyncTopLevelStatement [](http://example.com/codeflow?start=50&length=22)

Consider asserting the fact in order to avoid unintentional change to the scenario in the future.

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs:3963]: >NonAsyncTopLevelStatements [](http://example.com/codeflow?start=50&length=26)

Consider asserting the fact in order to avoid unintentional change to the scenario in the future.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 1)

**Comment by jcouv** [comment]
@dotnet/roslyn-compiler for second review. Thanks

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 3)

**Review by RikkiGibson — APPROVED** [review]
LGTM modulo additional tests discussed offline.

**Comment by jcouv** [comment]
While walking Rikki through the change, I realized a case was missed. Added tests and additional error code.
But this made me think of an implementation that wouldn't rely on filtering out some known error code. It turns out to be a good simplification.

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:9]: Is this still needed?

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:121]: >new BoundAwaitableValuePlaceholder(syntax, expression.Type); [](http://example.com/codeflow?start=30&length=60)

Consider inlining, To take even more advantage from an early return.

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:123]: >BindAwaitInfo [](http://example.com/codeflow?start=16&length=13)

It looks like we can call ```GetAwaitableExpressionInfo``` instead and avoid more unnecessary work

Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:123]: >expression [](http://example.com/codeflow?start=113&length=10)

It looks like passing ```expressionOpt``` is not necessary when we pass placeholder

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:123]: > do you think it would be better to call ```BindAwait``` and pass an additional flag to skip the async context check?

I think it is better to have logic that doesn't check content of a DiagnosticBag as long as the logic is sufficiently reliable

**Comment by AlekseyTs** [comment]
Done with review pass (commit 6)

File: src/Compilers/CSharp/Portable/Binder/Binder_Await.cs
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Binder/Binder_Await.cs:127]: >placeholder [](http://example.com/codeflow?start=58&length=11)

Could we simply pass ```expression``` here. It looks like this is what ```ReportBadAwaitContext``` would do in commit 6.

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 8)

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs:4025]: For some reason I thought we didn't make this suggestion because it's bad to await in a lock.

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/BindingAsyncTests.cs:4025]: For `async` method we consider it's always okay to await, regardless of context :-/ This isn't great, but it's an existing issue/question, orthogonal to what I'm fixing.

**Pull request merged**
